include/capypdf.hpp

```diff
@@ -169,6 +176,157 @@ class GraphicsState : public CapyC<CapyPDF_GraphicsState> {
     void set_TK(int32_t value) { CAPY_CPP_CHECK(capy_graphics_state_set_TK(*this, value)); }
 };
 
+class Function : public CapyC<CapyPDF_Function> {
+public:
+    friend class Generator;
+
+    Function(double *domain, int32_t num_domain, Color &c1, Color &c2, double n = 1.0) {
+        CapyPDF_Function *fn;
+        CAPY_CPP_CHECK(capy_type2_function_new(domain, num_domain, c1, c2, n, &fn))
+        _d.reset(fn);
+    }
+
+    Function(double *domain,
+             int32_t num_domain,
+             CapyPDF_FunctionId *functions,
+             int32_t num_functions,
+             double *bounds,
+             int32_t num_bounds,
+             double *encode,
+             int32_t num_encode) {
+        CapyPDF_Function *fn;
+        CAPY_CPP_CHECK(capy_type3_function_new(domain,
+                                               num_domain,
+                                               functions,
+                                               num_functions,
+                                               bounds,
+                                               num_bounds,
+                                               encode,
+                                               num_encode,
+                                               &fn))
+        _d.reset(fn);
+    }
+};
+
+class Shading : public CapyC<CapyPDF_Shading> {
+public:
+    friend class Generator;
+};
```

Comment:
The design goals of the C++ API have been to make things as simple as possible and to avoid inheritance whenever possible. In Python you can't really avoid inheritance, but in this you can. The various shadings have no functionality in common and they can't be used in place of each other. Thus they should be their own objects and Generator should have something along the lines of:

```
    CapyPDF_ShadingId add_shading(Shading2 &sh) {
        return add_shading_impl(sh);
    }
    CapyPDF_ShadingId add_shading(Shading6 &sh) {
        return add_shading_impl(sh);
    }
private:
    CapyPDF_ShadingId add_shading_impl(CapyPDF_Shading*sh) {
        CapyPDF_ShadingId sid;
        CAPY_CPP_CHECK(capy_generator_add_shading(*this, sh, &sid));
        return sid;
    }
```